builtin/logical/database/backend.go

```diff
@@ -286,3 +485,333 @@ cassandra, mssql, mysql, postgres
+				item.Priority = time.Now().Add(10 * time.Second).Unix()
+
+				// preserve the WALID if it was returned
+				if resp.WALID != "" {
+					item.Value = resp.WALID
+				}
+
+				if err := b.pushItem(item); err != nil {
+					b.logger.Warn("unable to push item on to queue", "error", err)
+				}
+				// go to next item
+				continue
+			}
+
+			lvr := resp.RotationTime
+			if lvr.IsZero() {
+				lvr = time.Now()
+			}
+
+			// update priority and push updated Item to the queue
+			nextRotation := lvr.Add(role.StaticAccount.RotationPeriod)
+			item.Priority = nextRotation.Unix()
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+		} else {
+			// highest priority item does not need rotation, so we push it back on
+			// the queue and break the loop
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+			break
+		}
+	}
+	return nil
+}
+
+// findStaticWAL loads a WAL entry by ID. If found, only return the WAL if it
+// is of type staticWALKey, otherwise return nil
+func (b *databaseBackend) findStaticWAL(ctx context.Context, s logical.Storage, id string) *setCredentialsWAL {
+	wal, err := framework.GetWAL(ctx, s, id)
+	if err != nil {
+		b.Logger().Warn("error reading WAL", "id", id, "error", err)
+		return nil
+	}
+
+	if wal == nil || wal.Kind != staticWALKey {
+		return nil
+	}
+
+	var walEntry setCredentialsWAL
+	if err = mapstructure.Decode(wal.Data, &walEntry); err != nil {
+		b.Logger().Warn("error decoding walEntry", err.Error())
+		return nil
+	}
+
+	return &walEntry
+}
+
+type setStaticAccountInput struct {
+	RoleName   string
+	Role       *roleEntry
+	Password   string
+	CreateUser bool
+	WALID      string
+}
+
+type setStaticAccountOutput struct {
+	RotationTime time.Time
+	Password     string
+	// Optional return field, in the event WAL was created and not destroyed
+	// during the operation
+	WALID string
+}
+
+// setStaticAccount sets the password for a static account associated with a
+// Role. This method does many things:
+// - verifies role exists and is in the allowed roles list
+// - loads an existing WAL entry if WALID input is given, otherwise creates a
+// new WAL entry
+// - gets a database connection
+// - accepts an input password, otherwise generates a new one via gRPC to the
+// database plugin
+// - sets new password for the static account
+// - uses WAL for ensuring passwords are not lost if storage to Vault fails
+//
+// This method does not perform any operations on the priority queue. Those
+// tasks must be handled outside of this method.
+func (b *databaseBackend) setStaticAccount(ctx context.Context, s logical.Storage, input *setStaticAccountInput) (*setStaticAccountOutput, error) {
+	// lvr is the known LastVaultRotation
+	var lvr time.Time
+	var merr error
+	// re-use WAL ID if present, otherwise PUT a new WAL
+	if input == nil {
+		input = &setStaticAccountInput{}
```

Comment:
If input is nil should we return, possibly with an error?

Reply:
Done in 29b056487